### DISCUSSION TITLE
knitr engins' option names dot `.` v.s. hyphen `-`

### DISCUSSION THREAD

**Discussion by eitsupi** [quarto-dev/quarto-cli#4120] [Q&A] [answered]
Related to #4107 and eitsupi/prqlr#58

Currently Quarto CLI has a built-in process for converting knitr option names.
Is there any chance that this process could be moved to the knitr side?
My understanding is that because this processing is on the Quarto CLI side, we cannot use option names like `engine-opts` in R Markdown. (if we set this, the option name will be changed to `engine.opts` when render via Quarto CLI)

I am working on implementing a new knitr engine in the prqlr package and am wondering if the option names should be in the dot pattern or the hyphen pattern.

**Comment by cderv** [comment, marked as answer]
This is a feature request for the **knitr** 📦 I believe. **knitr** option does not use `-` because it would need to be quoted when using in chunk header 
````
```{r, fig-cap="image"}
plot(cars)
```
````
would throw an error but 
````
```{r, 'fig-cap'="image"}
plot(cars)
```
````
would not.

I agree there is indeed something to think about though here, because we only convert known options.  

Regarding `engine.opts`, it is used extensively to pass options to some engine, often using CLI tool. It is a way to get option namespaced and avoid conflict too.

Happy to discuss this in an issue in **knitr** repo to see how we can improve for new engine author. 

Thanks for opening the discussion

**Reply by eitsupi** [reply]
Thank you for your response.
I finally chose to accept both patterns because I thought it would be confusing for users to have to explain that hyphens can only be used with Quarto CLI.

https://github.com/eitsupi/prqlr/blob/a4779c5ec296bf89650259305ec7601b351cedf6/R/knitr-engine.R#L43-L45
https://github.com/eitsupi/prqlr/blob/a4779c5ec296bf89650259305ec7601b351cedf6/vignettes/knitr.Rmd#L122-L156

````markdown
```{r}
#| echo: false

library(prqlr)
```

## YAML-style

```{prql}
#| engine-opts:
#|   dialect: mssql
#|   signature_comment: false

from mtcars
filter cyl > 6
select [mpg]
take 3
```

## R-style

```{prql engine.opts=list(dialect="mssql", signature_comment=FALSE)}
from mtcars
filter cyl > 6
select [mpg]
take 3
```
````

> This is a feature request for the **knitr** 📦 I believe.

My opinion is that automatic renaming of options on the Quarto side is not a good implementation that will confuse users.
On the other hand, I also think that using dots for names on YAML should be avoided given the existence of tools like jq, so I just wanted to say that this feature should be implemented in knitr instead of Quarto CLI.

**Reply by cderv** [reply]
Thanks I have open https://github.com/yihui/knitr/issues/2214 to continue discussion, track interest and think about that. 

> My opinion is that automatic renaming of options on the Quarto side is not a good implementation that will confuse users.

Quarto use `-` for option name by design for all options, including cells. The automatic conversion is by convenience to support **knitr** options that way instead of explaning `-` for quarto only option, and `.` for **knitr** options. So that it is less confusing... 😅

**Comment by cderv** [comment]
https://github.com/yihui/knitr/issues/2214 is now closed.

With future **knitr** 1.44 (from dev version 1.43.7) now any knitr options using a `-` will be converted to its `.` counterpart. This will happen for all documents, included in Quarto context. 

This means that `engine-opts` will work in R Markdown context too.
